File: src/main/java/no/nav/bidrag/dokument/arkiv/service/JournalpostService.java
package no.nav.bidrag.dokument.arkiv.service;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;
import no.nav.bidrag.dokument.arkiv.consumer.PersonConsumer;
import no.nav.bidrag.dokument.arkiv.consumer.SafConsumer;
import no.nav.bidrag.dokument.arkiv.dto.Bruker;
import no.nav.bidrag.dokument.arkiv.dto.BrukerType;
import no.nav.bidrag.dokument.arkiv.dto.Journalpost;
import no.nav.bidrag.dokument.arkiv.dto.PersonResponse;
import no.nav.bidrag.dokument.arkiv.dto.Sak;
import no.nav.bidrag.dokument.arkiv.dto.TilknyttetJournalpost;
import no.nav.bidrag.dokument.arkiv.model.PersonException;
import no.nav.bidrag.dokument.dto.JournalpostDto;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;

public class JournalpostService {

  private static final Logger LOGGER = LoggerFactory.getLogger(JournalpostService.class);

  private final SafConsumer safConsumer;
  private final PersonConsumer personConsumer;

  public JournalpostService(SafConsumer safConsumer, PersonConsumer personConsumer) {
    this.safConsumer = safConsumer;
    this.personConsumer = personConsumer;
  }

  public Optional<Journalpost> hentJournalpost(Long journalpostId) {
    return hentJournalpost(journalpostId, null);
  }

  public Optional<Journalpost> hentJournalpostMedFnrOgTilknyttedeSaker(Long journalpostId, String saksnummer) {
    var jpOptional = hentJournalpostMedFnr(journalpostId, saksnummer);
    if (jpOptional.isEmpty()){
      return jpOptional;
    }
    return Optional.of(populerMedTilknyttedeSaker(jpOptional.get()));
  }

  public Optional<Journalpost> hentJournalpostMedAktorId(Long journalpostId) {
    var journalpost = hentJournalpost(journalpostId);
    return journalpost.map(this::konverterFnrTilAktorId);
  }

  public List<JournalpostDto> finnJournalposter(String saksnummer, String fagomrade) {
    return finnJournalposterForSaksnummer(saksnummer, fagomrade).stream()
        .map((this::konverterAktoerIdTilFnr))
        .filter((jp)-> !(jp.getTilleggsopplysninger().isEndretTema() || jp.getTilleggsopplysninger().isNyDistribusjonBestilt()) )
        .map(Journalpost::tilJournalpostDto)
        .collect(Collectors.toList());
  }

  private Optional<Journalpost> hentJournalpost(Long journalpostId, String saksnummer) {
    var journalpost = safConsumer.hentJournalpost(journalpostId);

    if (Objects.isNull(journalpost) || journalpost.erIkkeTilknyttetSakNarOppgitt(saksnummer)) {
      return Optional.empty();
    }

    return Optional.of(journalpost);
  }

  protected List<TilknyttetJournalpost> hentTilknyttedeJournalposter(Journalpost journalpost){
    if (journalpost.getDokumenter().isEmpty() || journalpost.getSak() == null){
      return List.of();
    }
    var dokumentInfoId = journalpost.getDokumenter().get(0).getDokumentInfoId();
    LOGGER.info(String.format("Henter tilknyttede journalposter for journalpost %s med dokumentinfoId %s", journalpost.getJournalpostId(), dokumentInfoId));
    return Optional.ofNullable(dokumentInfoId).map(safConsumer::finnTilknyttedeJournalposter).orElse(new ArrayList<>());
  }

  public Journalpost populerMedTilknyttedeSaker(Journalpost journalpost){
    var journalpostFagsakId = journalpost.getSak() != null ? journalpost.getSak().getFagsakId() : "";
    var tilknytteteJournalposter = hentTilknyttedeJournalposter(journalpost);
    var saker = tilknytteteJournalposter.stream()
        .map(TilknyttetJournalpost::getSak)
        .filter(Objects::nonNull)
        .map(Sak::getFagsakId)
        .filter(Objects::nonNull)
        .filter(fagsakId -> !Objects.equals(fagsakId, journalpostFagsakId)).toList();
    var sakerNoDuplicates = new HashSet<>(saker).stream().toList();
    LOGGER.info("Fant {} saker for journalpost {}", saker.size() + 1, journalpost.getJournalpostId());
    journalpost.setTilknyttedeSaker(sakerNoDuplicates);
    return journalpost;
  }

  private Optional<Journalpost> hentJournalpostMedFnr(Long journalpostId, String saksummer) {
    var journalpost = hentJournalpost(journalpostId, saksummer);
    return journalpost.map(this::konverterAktoerIdTilFnr);
  }

  public List<Journalpost> finnJournalposterForSaksnummer(String saksnummer, String fagomrade) {
    return safConsumer.finnJournalposter(saksnummer, fagomrade);
  }

  private Journalpost konverterAktoerIdTilFnr(Journalpost journalpost) {
    var bruker = journalpost.getBruker();
    if (Objects.isNull(bruker) || !journalpost.getBruker().isAktoerId()) {
      return journalpost;
    }

    var personResponse = hentPerson(bruker.getId());
    var brukerId = personResponse.getIdent();
    journalpost.setBruker(new Bruker(brukerId, BrukerType.FNR.name()));
    return journalpost;
  }

  private Journalpost konverterFnrTilAktorId(Journalpost journalpost) {
    var bruker = journalpost.getBruker();
    if (Objects.isNull(bruker) || journalpost.getBruker().isAktoerId()) {
      return journalpost;
    }

    var personResponse = hentPerson(bruker.getId());
    var brukerId = personResponse.getAktoerId();
    journalpost.setBruker(new Bruker(brukerId, BrukerType.AKTOERID.name()));
    return journalpost;
  }

  private PersonResponse hentPerson(String personId) {
    var personResponse = personConsumer.hentPerson(personId);
    if (!personResponse.is2xxSuccessful()) {
      throw new PersonException("Det skjedde en feil ved henting av person", HttpStatus.INTERNAL_SERVER_ERROR);
    }
    var response =  personResponse.getResponseEntity().getBody();
    if (Objects.isNull(response)){
      LOGGER.error("Fant ingen person med id {}", personId);
      return new PersonResponse(personId, personId);
    }
    return response;
  }
}

File: src/main/java/no/nav/bidrag/dokument/arkiv/service/JournalpostService.kt
package no.nav.bidrag.dokument.arkiv.service

import no.nav.bidrag.dokument.arkiv.consumer.PersonConsumer
import no.nav.bidrag.dokument.arkiv.consumer.SafConsumer
import no.nav.bidrag.dokument.arkiv.dto.Bruker
import no.nav.bidrag.dokument.arkiv.dto.BrukerType
import no.nav.bidrag.dokument.arkiv.dto.DistribusjonsInfo
import no.nav.bidrag.dokument.arkiv.dto.Journalpost
import no.nav.bidrag.dokument.arkiv.dto.Sak
import no.nav.bidrag.dokument.arkiv.dto.TilknyttetJournalpost
import no.nav.bidrag.dokument.arkiv.model.kanIkkeHenteJournalMedUgyldigFagomrade
import no.nav.bidrag.transport.dokument.JournalpostDto
import no.nav.bidrag.transport.person.PersonDto
import org.slf4j.LoggerFactory
import java.util.Optional

class JournalpostService(
    private val safConsumer: SafConsumer,
    private val personConsumer: PersonConsumer,
) {
    fun hentJournalpost(journalpostId: Long): Journalpost? = hentJournalpost(journalpostId, null)

    fun hentJournalpostMedTilknyttedeSaker(journalpostId: Long): Journalpost? = hentJournalpost(journalpostId)?.let { populerMedTilknyttedeSaker(it) }

    fun hentJournalpostMedFnrOgTilknyttedeSaker(journalpostId: Long, saksnummer: String?): Optional<Journalpost> = hentJournalpostMedFnr(
        journalpostId,
        saksnummer,
    )?.let {
        Optional.ofNullable(
            populerMedTilknyttedeSaker(it),
        )
    } ?: Optional.empty()

    fun List<String>.inneholderBidragFagomrader() = this.isEmpty() || this.hentIkkeBidragFagomrader().isEmpty()

    fun List<String>.hentIkkeBidragFagomrader() = this.filter { it != "BID" && it != "FAR" }
    fun finnJournalposter(saksnummer: String, fagomrade: List<String> = emptyList()): List<JournalpostDto> {
        if (!fagomrade.inneholderBidragFagomrader()) {
            kanIkkeHenteJournalMedUgyldigFagomrade(
                fagomrade.hentIkkeBidragFagomrader().joinToString(","),
            )
        }
        return finnJournalposterForSaksnummer(saksnummer, fagomrade)
            .map { konverterAktoerIdTilFnr(it) }
            .filter { !(it.tilleggsopplysninger.isEndretTema() || it.tilleggsopplysninger.isNyDistribusjonBestilt()) }
            .map { it.tilJournalpostDto() }
    }

    private fun hentJournalpost(journalpostId: Long, saksnummer: String?): Journalpost? {
        val journalpost = safConsumer.hentJournalpost(journalpostId)
        return if (journalpost.erIkkeTilknyttetSakNarOppgitt(saksnummer)) null else journalpost
    }

    fun hentTilknyttedeJournalposter(journalpost: Journalpost): List<TilknyttetJournalpost> {
        if (journalpost.dokumenter.isEmpty() || journalpost.sak == null) {
            return listOf()
        }
        val dokumentInfoId = journalpost.dokumenter[0].dokumentInfoId
        LOGGER.info("Henter tilknyttede journalposter for journalpost ${journalpost.journalpostId} med dokumentinfoId $dokumentInfoId")
        return dokumentInfoId?.let { finnTilknyttedeJournalposter(it) } ?: emptyList()
    }

    fun finnTilknyttedeJournalposter(dokumentreferanse: String): List<TilknyttetJournalpost> = safConsumer.finnTilknyttedeJournalposter(
        dokumentreferanse,
    )

    fun populerMedTilknyttedeSaker(journalpost: Journalpost): Journalpost {
        val journalpostFagsakId = if (journalpost.sak != null) journalpost.sak!!.fagsakId else ""
        val tilknytteteJournalposter = hentTilknyttedeJournalposter(journalpost)
        val saker = tilknytteteJournalposter
            .asSequence()
            .filter { it.isNotFeilregistrert() }
            .map(TilknyttetJournalpost::sak)
            .filterNotNull()
            .map(Sak::fagsakId)
            .filterNotNull()
            .filter { it != journalpostFagsakId }.toList()
        val sakerNoDuplicates = HashSet(saker).stream().toList()
        LOGGER.info("Fant ${saker.size + 1} saker for journalpost ${journalpost.journalpostId}")
        journalpost.tilknyttedeSaker = sakerNoDuplicates
        return journalpost
    }

    fun hentJournalpostMedFnr(journalpostId: Long, saksummer: String?): Journalpost? = hentJournalpost(journalpostId, saksummer)?.let {
        konverterAktoerIdTilFnr(it)
    }

    fun finnJournalposterForSaksnummer(saksnummer: String, fagomrade: List<String> = emptyList()): List<Journalpost> = safConsumer.finnJournalposter(
        saksnummer,
        fagomrade,
    )

    fun hentDistribusjonsInfo(journalpostId: Long): DistribusjonsInfo = safConsumer.hentDistribusjonInfo(journalpostId)

    private fun konverterAktoerIdTilFnr(journalpost: Journalpost): Journalpost {
        val bruker = journalpost.bruker
        if (bruker == null || !bruker.isAktoerId()) {
            return journalpost
        }
        personConsumer.hentPerson(bruker.id).ifPresent { (brukerId): PersonDto ->
            journalpost.bruker = Bruker(brukerId.verdi, BrukerType.FNR.name)
        }
        return journalpost
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(JournalpostService::class.java)
    }
}
